cip/1.accepted/CIP2021-08-10-Operator-precedence.adoc

```diff
@@ -0,0 +1,727 @@
+For instance, the following operator tree
+----
+    *
+   / \
+  +   4
+ / \
+3   2
+----
+encodes that the numeric addition is evaluated before the numeric multiplication, so that the expression results in `20`.
+
+However, the query string is a sequence of characters and not a tree.
+The parser turns the query string into such an operator tree.
+In the query string, nesting can clearly be denoted by parentheses.
+For instance, the query string
+----
+((3+2)*4)
+----
+allows a parser to unambiguously construct the operator tree shown above.
+
+For user convenience, better readability, and familiarity with common conventions in mathematical notation, openCypher (and most other query and programming languages) allows omitting the parentheses.
+For instance,
+----
+3+2*4
+----
+is also a valid expression.
+
+Without additional rules, however, it is not possible to unambiguously construct a operator tree for such an expression.
+In the particular case, two operator trees are possible:
+
+- `((3+2)*4)`
+- `(3+(2*4))`
+
+To avoid this kind of ambiguity, openCypher (and most other query and programming languages) has _precedence rules_.
+
+Precedence rules (or _rules of operator precedence_) define the order in which the different operators, i.e. the different kinds of expressions, are evaluated.
+For instance, numeric multiplication is evaluated before numeric addition, such that expression
+----
+3+2*4
+----
+is effectively evaluated as
+----
+(3+(2*4))
+----
+and, hence, results in `11` (rather than `20`).
+
+This CIP states the precedence rules of openCypher.
+
+== Proposal
+
+The precedence in openCypher is defined by (1) _precedence levels_ and (2) _operator associativity_.
+Both are defined in the following two subsections.
+
+=== Precedence levels
+
+The following <<precedenceLevels,table>> lists the precedence levels in descending order.
+
+.[[precedenceLevels]]Precedence levels
+[cols="<.<1a,<.<4a,<.<4a", options="header"]
+|===
+|Level         |Group         |Operators
+
+|12
+|Atoms
+|
+
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Literal[Literals]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Variable[Variables]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Parameter[Parameters]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=CaseExpression[Case expressions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ListComprehension[List comprehensions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=PatternComprehension[Pattern comprehensions]
+* https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Reduce[Reduce operator]
+* quantifiers
```

Comment:
Quantifiers do not have their own grammar production has they are not really referencable like the other things. I can 

a. Just extend the text here a bit, or
b. Modify the grammar so that they have their own production

The latter is the bigger change, but I still lean towards it, because it is a bit of an oddity that the grammar does not separate the quantifiers out into their own production. Opinions?

Reply:
I don't have a strong opinion, as long as we clarify what those quantifiers are. For what it's worth, Neo4j does the latter.

Reply:
Done. Note that link does not work yet, since it always goes the master branch.